[클린코드 리액트 2기 이현정] 페이먼츠 미션 Step 1

jlee0505 opened the pull request:
안녕하세요 동희 리뷰어님! 이번 페이먼츠 미션 step.1을 함께하게 된 리뷰이 이현정입니다. 잘 부탁드립니다! :)
---
## DONE ✅
    
### 카드 추가

- [x] <(뒤로가기) 버튼 클릭 시, 카드 목록 페이지로 이동한다.
- [x] 카드 번호를 입력 받을 수 있다.

  - [x] 카드 번호는 숫자만 입력가능하다.
  - [x] 카드 번호 4자리마다 -가 삽입된다.
  - [x] 카드 번호는 실시간으로 카드 UI에 반영된다.
  - [x] 카드 번호는 앞 8자리만 숫자로 보여지고, 나머지 숫자는 \*로 보여진다.
  - [x] [추가] 유효하지 않은 카드 번호를 입력하면, 입력 폼 아래에 에러 메시지를 보여준다.
  - [x] [추가] 4자리를 다 채우면 다음 번호로 커서가 옮겨간다.

- [ ] 만료일을 입력 받을 수 있다.

  - [x] MM / YY 로 placeholder를 적용한다.
  - [x] 월, 년 사이에 자동으로 /가 삽입된다.
  - [x] 만료일은 실시간으로 카드 UI에 반영된다.
  - [x] 월은 1이상 12이하 숫자여야 한다.
  - [x] [추가] 년도는 현재년도보다 과거일 수 없다.
  - [x] [추가] 월 입력 길이가 2이상이면, 년도 입력으로 focus된다.
  - [ ] [추가] 월, 년 입력이 유효하면 보안코드 입력으로 focus된다.
  - [x] [추가] 유효하지 않은 만료일을 입력하면, 입력 폼 아래에 에러 메시지를 보여준다.

- [x] 카드 소유자 이름을 입력 받을 수 있다.

  - [x] 이름은 30자리까지 입력할 수 있다.
  - [x] 이름 입력 폼 위에, 현재 입력 자릿수와 최대 입력 자릿수를 실시간으로 보여준다.

- [ ] 보안코드를 입력 받을 수 있다.

  - [x] 보안코드는 \*으로 보여진다.
  - [x] 보안코드는 숫자만 입력가능하다.
  - [x] [추가] 유효하지 않은 보안코드를 입력하면, 입력 폼 아래에 에러 메시지를 보여준다.
  - [ ] [추가] 보안코드 3자리가 입력되면 카드 비밀번호 입력으로 focus된다.

- [x] 카드 비밀번호의 앞 2자리를 입력 받을 수 있다.

  - [x] 카드 비밀번호는 각 폼마다 한자리 숫자만 입력가능하다.
  - [x] 카드 번호 입력 시, \*으로 보여진다.
  - [x] [추가] 첫자리 입력이 완료되면 둘째자리 입력으로 focus된다.
  - [x] [추가] 유효하지 않은 카드 비밀번호를 입력하면, 입력 폼 아래에 에러 메시지를 보여준다.

- [x] 입력이 완료되면, 카드 등록 완료 페이지로 이동한다.

  - [x] [추가] 카드 비밀번호의 마지막자리가 입력 완료되면, form이 submit 된다.
  - [x] [추가] submit되면, 빈칸 없이 모두 입력되었는지 체크한다.
  - [x] [추가] 빈칸이 남아있다면 알림창을 띄운다.
  - [x] [추가] 빈칸 없이 모두 채워졌다면, 카드 등록 완료 페이지로 이동한다.

- [ ] [추가] 입력이 완료되면, 모달이 뜬다.

  - [ ] [추가] 빈칸이 남아있다면 비어있는 빈칸 중 첫번째 빈칸으로 focus 한다.
  - [ ] [추가] 빈칸 없이 모두 채워졌다면, 모달이 뜬다.

- [ ] [추가] 모달 중 회사를 선택할 수 있다.

  - [ ] [추가] 카드 회사를 선택하면, 카드 배경색이 변한 후 모달이 사라진다.
  - [ ] [추가] 모달이 뜨는 페이지는 하단에 '다음' 버튼이 존재한다.
  - [ ] [추가] '다음' 버튼을 누르면 카드 목록 페이지로 이동한다.
  - [ ] [추가] 카드 클릭시 카드 회사를 재선택할 수 있다.

### To-do

- [ ] [fix] 패스워드 마지막 한자리 submit 안되는 오류 해결

## 리뷰어님께 🙋🏻‍♀️

 ### 이런 점을 고민했어요
  - 데이터 스키마: 어떤 데이터가 어디에 있어야 하는 걸까
  - Storybook 을 이용한 CDD: 컴포넌트 위주로 개발하면 어떤 점이 좋을까
 
### 이렇게 진행했어요
- **CDD 시도:  선 UI 정적 컴포넌트 제작(with Storybook) 후 기능 구현**
    - [[질문](https://github.com/next-step/react-payments/compare/%EC%84%A4%EA%B3%84)](설계) 전반적인 진행 방향은 [[공식 문서](https://ko.reactjs.org/docs/thinking-in-react.html)](https://ko.reactjs.org/docs/thinking-in-react.html)를 참고해서 해보았는데, 제대로 적용한 게 맞는지 이번 미션에서 의도한 CDD가 맞는지 궁금합니다!
    - [[질문](https://github.com/next-step/react-payments/compare/%EC%84%A4%EA%B3%84)](설계) 스토리북으로 테스트를 한다는 개념이 잘 와닿지 않습니다. UI테스트만 하는 건지, 기능 구현도 함께 하는 건지 헷갈립니다. UI뿐 아니라 기능 구현에 있어서도 스토리북으로 테스트를 진행했어야 하는걸까요?
    - [[질문](https://github.com/next-step/react-payments/compare/%EC%84%A4%EA%B3%84)](설계) 데이터 스키마라는 개념을 이번 미션으로 공부하며 접하게 되었는데요, UI와 데이터 구조는 비슷하기 때문에 UI설계를 잘하는 것과 데이터를 잘 나누는 것은 연관이 있다고 배웠습니다. 컴포넌트 분리를 잘 한 건지 확신이 서지 않는데 이부분에 대해서 리뷰어님의 의견이 궁금합니다!
    - [[질문](https://github.com/next-step/react-payments/compare/%ED%81%B4%EB%A6%B0%EC%BD%94%EB%93%9C)](클린코드) 진행하면서 반복되는 로직이 많다고 느꼈는데 클린코드 관점에서의 리뷰어님의 의견이 궁금합니다!
  
- **덧) 추가 미션**
    
    미션을 진행하면서 ‘이런 기능을 추가하면 좋지 않을까?’ 생각이 드는 기능에 [추가] 타이틀을 붙이고 시도해보았습니다. REQUIREMENT.md에 [추가]가 붙어있지 않은 기능들이 미션-step.1 에서 주어진 기본 기능들입니다. 
    
    추가기능까지 구현하려다보니 조잡해진 부분이 있는데, 욕심을 버리고 우선 한 것 까지만 먼저 피드백을 받고자 합니다. 차후 계속 디벨롭 예정입니다!
---
감사합니다 🙇🏻‍♀️

jlee0505 committed:
chore: init project

- cra 설치
- storybook 설치
- 페이지 기반 라우팅
- 컴포넌트 분리 골격 잡기

jlee0505 committed:
chore: storybook을 이용한 UI컴포넌트(정적버전) 구축

jlee0505 committed:
chore: 누락된 Button UI 컴포넌트 추가

jlee0505 committed:
chore: Header UI 컴포넌트 스타일 변경

jlee0505 committed:
chore: class -> className 변경

jlee0505 committed:
chore: UI컴포넌트 구조 수정

- 기존 RegistrationPage에서 CardNicknamePage 분리

jlee0505 committed:
chore: 마이너 스타일 변경

jlee0505 committed:
docs: step.1 기능 구현을 위한 구상 및 기능 목록 수정

jlee0505 committed:
feat: 뒤로 가기 기능 추가

jlee0505 committed:
feat: 카드 번호 입력 기능 추가

- CardForm 아래 input들은 모두 캡슐화했습니다.
-- 이유: CardRegistrationPage를 Header, Card, CardForm 으로 선언적 명시하기 위해서

jlee0505 committed:
feat: 만료일 입력 기능 추가

jlee0505 committed:
feat: 카드 소유자 입력 기능 추가

jlee0505 committed:
feat: 보안코드(CVC) 입력 기능 추가

jlee0505 committed:
feat: 비밀번호 입력 기능 추가

jlee0505 committed:
feat: 데이터 전역관리를 위한 store 생성

jlee0505 committed:
fix: 컨텍스트 에러 고침

jlee0505 committed:
chore: 상수화

jlee0505 committed:
feat: 입력을 마치면 카드 등록 완료 페이지로 이동

jlee0505 committed:
Merge Version-1 with Main Branch (#2)

* chore: init project

- cra 설치
- storybook 설치
- 페이지 기반 라우팅
- 컴포넌트 분리 골격 잡기

* chore: storybook을 이용한 UI컴포넌트(정적버전) 구축

* chore: 누락된 Button UI 컴포넌트 추가

* chore: Header UI 컴포넌트 스타일 변경

* chore: class -> className 변경

* chore: UI컴포넌트 구조 수정

- 기존 RegistrationPage에서 CardNicknamePage 분리

* chore: 마이너 스타일 변경

* docs: step.1 기능 구현을 위한 구상 및 기능 목록 수정

* feat: 뒤로 가기 기능 추가

* feat: 카드 번호 입력 기능 추가

- CardForm 아래 input들은 모두 캡슐화했습니다.
-- 이유: CardRegistrationPage를 Header, Card, CardForm 으로 선언적 명시하기 위해서

* feat: 만료일 입력 기능 추가

* feat: 카드 소유자 입력 기능 추가

* feat: 보안코드(CVC) 입력 기능 추가

* feat: 비밀번호 입력 기능 추가

* feat: 데이터 전역관리를 위한 store 생성

* fix: 컨텍스트 에러 고침

* chore: 상수화

* feat: 입력을 마치면 카드 등록 완료 페이지로 이동

bigsaigon333 commented on react-payments-app/src/components/Card/CardShapedButton/card.jsx:
```diff
@@ -0,0 +1,11 @@
+const CardShapedButton = () => {
+  return (
+    <div class="card-box">
+      <div class="empty-card">+</div>
```
`npm start` 로 실행해서 보니 콘솔창에 아래와 같은 에러가 뜨네요. 
왜 뜨는걸까요? 😄 

![image](https://user-images.githubusercontent.com/31029000/219868665-0243925e-6268-47d9-9c37-1cacd0d66f26.png)

jlee0505 replied:
앗 수정하였습니다! [bcabb2d](https://github.com/next-step/react-payments/pull/43/commits/bcabb2d246f435a0029d23b9e3910bd521a958f0)

warning도 꼼꼼히 잡는 습관을 들여야 겠어요. 감사합니다!

bigsaigon333 commented on react-payments-app/.storybook/main.js:
```diff
@@ -0,0 +1,16 @@
+  "core": {
+    "builder": "@storybook/builder-webpack5"
+  }
+}
```
요 마크를 왜 깃헙은 보여주는 걸까요?

어떻게 하면 없앨 수 있을까요?

자동화 할 수 있는 방법은 없을까요?

![image](https://user-images.githubusercontent.com/31029000/219869623-a0f2f770-04b0-4abb-863f-174cec75ac44.png)

bigsaigon333 commented on react-payments-app/public/index.html:
```diff
@@ -0,0 +1,15 @@
+</head>
+<body>
+  <div id="root">
+    <h2 align="middle">페이먼츠</h2>
+    <p align="middle">React 모바일 페이먼츠 애플리케이션</p>
```
요거는 왜 넣으신걸까요..?
요기 실제 렌더링 되는 html에 포함이 될까요?
align 속성을 꼭 사용해야 할까요?

![image](https://user-images.githubusercontent.com/31029000/219869776-e0e0caad-702e-43d7-a8a2-572ec1b814e2.png)

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/p#attributes

jlee0505 replied:
초기에 컴포넌트 분리하면서 실수로 남겨둔 것 같아요....삭제하였습니다!

bigsaigon333 commented on react-payments-app/src/components/Card/Card.jsx:
```diff
@@ -0,0 +1,36 @@
+    <div id="card">
+      <div className="flex-column-center mb-10">
+        <div className="card-box">
+          <div className={cardStatus}>
```
`cardStatus`가 `className`에 바로 할당되는 건 안전하지 않은 방법인 것 같아요. 

아래와 같이 존재하지 않는 className이 cardStatus로 주어지면, 바로 스타일링이 깨져요 😭 

![image](https://user-images.githubusercontent.com/31029000/219870648-166503f6-345d-417a-9f48-26cd1b108c0d.png)

bigsaigon333 replied:
위에 얘기와 별개로, 스토리북에서는 props를 특정 값으로만 제한하게 하는 기능이 있습니다 😄 
(https://storybook.js.org/addons/storybook-addon-knobs-color-options)

여력이 되시면 한번 적용해보는 것도 좋을거 같아요 👍🏻

bigsaigon333 commented on react-payments-app/src/components/Card/CardShapedButton/card.stories.jsx:
```diff
@@ -0,0 +1,11 @@
+
+const Template = (args) => <CardShapedButton {...args} />;
+
+export const Default = Template.bind({});
```
```suggestion
import CardShapedButtonComponent from "./card";
import "../card.css";

export default {
  title: "Components/Card/CardShapedButton",
  component: CardShapedButtonComponent,
};

const Template = (args) => <CardShapedButton {...args} />;

export const CardShapedButton = Template.bind({});
```

(소소)

오직 하나의 스토리만 있고, 스토리명이 컴포넌트명과 일치한다면 아래와 같이 depth 없이 바로 나타낼 수 있습니다 😄 

![image](https://user-images.githubusercontent.com/31029000/219870947-e746eb0f-82e8-4c48-9163-f6cdaea639b5.png)


https://storybook.js.org/docs/react/writing-stories/naming-components-and-hierarchy#single-story-hoisting

jlee0505 replied:
오...! 꿀팁 감사합니다 😁

bigsaigon333 commented on react-payments-app/src/components/Header/Header.jsx:
```diff
@@ -0,0 +1,22 @@
+          >
+            {headerIcon}
+          </div>
+        )}
```
`headerIcon`에 `0`이라는 값이 들어오면 무엇이 렌더링될까요?

JS에서는 props로 어떤 값이든 다 올 수 있으므로, 이런 표현 사용시 각별히 조심해야 합니다. 

(TS를 쓴다면 어느정도 방어는 되지만, 그래도 비슷한 문제는 존재합니다)

bigsaigon333 commented on react-payments-app/src/components/Header/Header.jsx:
```diff
@@ -0,0 +1,22 @@
+const Header = ({ pageTitle, headerIcon, onClick }) => {
+  return (
+    <>
```
Fragment가 꼭 필요할까요?

bigsaigon333 commented on react-payments-app/src/components/Header/Header.jsx:
```diff
@@ -0,0 +1,22 @@
+            className="cursor-pointer"
+            onClick={(e) => {
+              onClick(e);
+            }}
```
이렇게 쓰신 이유가 있으실까요? 

```suggestion
            onClick={onClick}
```

위와 같이 써도 충분할거 같아서요 😄

bigsaigon333 commented on react-payments-app/src/components/Header/Header.jsx:
```diff
@@ -0,0 +1,22 @@
+      </div>
+    </>
+  );
+};
```
사실 이런 Header 컴포넌트는 생각보다 변경에 매우 취약합니다.

`headerIcon` 을 왼쪽이 아니라 오른쪽에 위치하게 하고 싶다면 어떻게 할 수 있을까요?

왼쪽에 둘지 오른쪽에 둘지를 props로 받으면 될까요?

pageTitle 하단에 작은 pageDescription을 추가하고 싶다면 어떻게 할 수 있을까요?



이 Header 컴포넌트의 역할은 정확히 무엇일까요? (무엇이길 기대하시나요?)

일단은 물음표만 던져보겠습니다 😄 

계속해서 논의해봐요~

bigsaigon333 commented on react-payments-app/src/components/Header/header.css:
```diff
@@ -0,0 +1,9 @@
+.page-title {
+  font-weight: 500;
+  font-size: 20px;
+  line-height: 22px;
+  display: flex;
+  align-items: center;
+
+  color: #383838;
+}
```
`button.css` 에도 동일한 클래스명의 `.page-title`이 있다면 어떻게 될까요? 😃

jlee0505 replied:
음... button도 header와 같은 스타일 효과를 주는 걸 의도하신걸까요?

bigsaigon333 replied:
음.. 동일한 클래스명이 중복되면 어떤 문제가 발생할까? 라는 질문이었습니다~!

bigsaigon333 commented on react-payments-app/src/components/Input/CardExpirationDate/input.jsx:
```diff
@@ -0,0 +1,74 @@
+import { useEffect, useState, useRef } from "react";
+import { MAX_INPUT_LENGTH, VALID_INPUT } from "../../../common/constant";
+
+const CURRENT_YEAR = Number(new Date().getFullYear().toString().slice(-2));
```
```suggestion
const CURRENT_YEAR = new Date().getFullYear() % 100
```

이렇게 할 수도 있습니다 😄

bigsaigon333 replied:
`CURRENT_YEAR`를 두자리수로 변환해서 계산하는게 좋을까요, 비교할 `expirationDate[1]` 를 네자리수로 변환해서 계산하는게 좋을까요?

현정님의 생각이 궁금하네요 😸

bigsaigon333 commented on react-payments-app/src/components/Input/CardForm.jsx:
```diff
@@ -0,0 +1,70 @@
+  };
+
+  return (
+    <>
```
이 Fragment도 꼭 필요한 걸까요?

jlee0505 replied:
그렇네요! 필요없는 <></> 를 삭제해주었습니다!

bigsaigon333 commented on react-payments-app/src/components/Input/CardForm.jsx:
```diff
@@ -0,0 +1,70 @@
+  const handleSubmit = (e) => {
+    e.preventDefault();
+
+    console.log(password);
```
```suggestion
```

프로덕션 코드에 console.log로 정보를 남기는건 좋지 않은거 같아요 😅

bigsaigon333 commented on react-payments-app/src/components/Input/input.css:
```diff
@@ -34,7 +43,12 @@
+}
+.input-bigger-text {
+  font-size: 150%;
+}
```
```suggestion

.input-of-pw {
  background: none;
}

.input-bigger-text {
  font-size: 150%;
}

```

클래스와 클래스 사이는 개행문자를 넣어주면 가독성이 좋아집니다 😄

bigsaigon333 commented on react-payments-app/src/store/CardContext.jsx:
```diff
@@ -0,0 +1,25 @@
+
+  if (!context) {
+    throw new Error("Need to be within CardContext to use useCard.");
+  }
```
👍🏻 👍🏻 👍🏻

bigsaigon333 commented on react-payments-app/src/pages/CardRegistrationPage/index.stories.jsx:
```diff
@@ -0,0 +1,17 @@
+const Template = (args) => <CardRegistrationPage {...args} />;
+
+export const BeforeRegistration = Template.bind({});
+BeforeRegistration.args = {};
+
+export const CompanySelection = Template.bind({});
+CompanySelection.args = {};
+
+export const AfterRegistration = Template.bind({});
+AfterRegistration.args = {};
```
`args` 에 값을 넣지 않는다면 굳이 스토리를 3개나 만들필요가 없지 않을까요? 

스토리파일은 마치 주석과도 같아서, 꾸준히 관리하지 않으면 금방 실제 코드와의 괴리가 발생합니다.

모든 컴포넌트에 스토리를 추가하는게 일정상의 이유로 어렵다면, 제대로 동작하지 않는 스토리는 일단 지우는것도 방법입니다..!

jlee0505 replied:
오...! 스토리파일은 주석 같다는 점이 인상 깊어요!
이부분은 추후 카드회사선택 모달 기능을 생각하며 짜두었어요. 주석의 TODO 느낌으로요!
이런 경우 스토리를 삭제했다가 나중에 작성하는 게 더 바람직할까요? 리뷰어님의 의견이 궁금합니다.

bigsaigon333 replied:
이미 만들었다면 지울 필요는 없을거 같아요..!

저희는 학습목적으로 미션을 진행하기 때문에, Step1 , Step2, Step3 가 이어진다는 걸 알지만,
실제 현업에서는 하나의 PR이 하나의 독립적인 기능의 종결을 의미하기 때문에, 
저는 스토리를 남기지 않을거  같아요! 아니면 `// TODO: ...` 과 같은 주석을 추가할수도 있을거 같구요~!

bigsaigon333 commented on react-payments-app/src/pages/CardListPage/index.jsx:
```diff
@@ -0,0 +1,22 @@
+  );
+};
+
+export default CardListPage;
```
CardListPage 에서 + 버튼을 눌러도 변화가 없어서, 매우 당황했습니다 💦

코드를 살펴보니 `/registration` 라우트로 제가 직접 입력해서 들어가야 볼 수 있게끔 구현되어 있네요.

FE개발자는 사용자와의 접점이 매우 중요합니다. 사용자와의 접점을 잘, 효율적으로 하기 위해 좋은 설계와 좋은 코드가 필요한거구요. 그 점을 염두에 두시고 개발하셨으면 좋겠습니다 😄

bigsaigon333 commented on react-payments-app/src/App.jsx:
```diff
@@ -0,0 +1,21 @@
+import CardNicknamePage from "./pages/CardNicknamePage";
+import CardRegistration from "./pages/CardRegistrationPage";
+
+function App() {
```
어떤 컴포넌트는 화살표 함수로, 어떤 컴포넌트는 function 함수로 작성하셨네요..!

어떤 기준으로 나누신건지 궁금합니다 😄

jlee0505 replied:
앗 이것은...cra 설치시 자동으로 이렇게 설치가 된 걸 제가 수정을 깜빡했네요😅 저는 화살표함수를 선호하기에 이부분도 그렇게 바꿔주었습니다.

bigsaigon333 commented on react-payments-app/src/index.jsx:
```diff
@@ -0,0 +1,18 @@
+import React from "react";
+import ReactDOM from "react-dom/client";
```
(소소)
다른 코드에서는 `import { createContext, useContext, useState } from "react";` 요렇게 각각을 import 하셨는데,
여기서는 default를 `React`로 import 하셨는데, 어떤 기준인지 궁금합니다..!

jlee0505 replied:
오...이부분 역시 cra를 설치하면 자동으로 되어있던 부분이라 신경쓴 적이 없었어요. 같은 기준으로 import를 하는 게 깔끔할 듯 하여 import { StrictMode } from 'react'; 로 수정해 주었습니다!

bigsaigon333 commented on react-payments-app/src/pages/CardRegistrationPage/index.jsx:
```diff
@@ -0,0 +1,59 @@
+import { useState } from "react";
```
사용하지 않는 import들과, 변수들이 많은 거 같아요..!

특히나 구조분해한 변수를 사용하지 않는건, 
지우는걸 그냥 깜빡한건지, 아니면 해당 변수를 사용해야하는데 사용하지 않은 버그인건지 구별이 안가기 때문에 
코드의 신뢰도를 떨어트리는 거 같아요..!

lint 사용을 생활화 하시는걸 추천드립니다 👍🏻 👍🏻

bigsaigon333 commented on react-payments-app/src/components/Input/CardPassword/input.jsx:
```diff
@@ -0,0 +1,70 @@
+    }
+  };
+
+  useEffect(() => {
+    onChange(password, error);
+    if (password[1]) {
+      console.log(password[1]);
+      handleSubmit(event);
+    }
+  }, [password, error]);
```
`useEffect`를 이용해 상태를 동기화 시키는 건 추천되는 방법은 아닙니다.

그냥 password와 error가 바뀌는 곳에서 onChange를 호출하는 방법이 더 낫습니다 😄

jlee0505 replied:
오 그렇게도 할 수 있겠네요! 그런데 어떤 이유에서 useEffect 사용이 지양되는 걸까요?

bigsaigon333 replied:
https://beta.reactjs.org/learn/you-might-not-need-an-effect# 정독을 추천드립니다~!

간단히 말해보자면, 
- 렌더 페이즈에서 React는 변경사항을 DOM에 커밋하여 화면을 업데이트하고 그 다음에 Effect를 실행하는데요, Effect에서 상태 변경이 있다면 또다시 변경사항을 DOM에 커밋하여 화면을 업데이트하는 작업을 처음부터 다시 실행합니다. -> 렌더 페이즈에서 상태 변경을 할 것이 추천되고 있씁니다~(갑론을박이 좀 있긴 합니다)
- Effect는 외부 시스템과 상태를 싱크하기 위해서 사용되어야 합니다.

bigsaigon333 commented on react-payments-app/src/components/Input/CardPassword/input.jsx:
```diff
@@ -0,0 +1,70 @@
+    if (password[1]) {
+      console.log(password[1]);
+      handleSubmit(event);
+    }
```
```suggestion
    if (password[1]) {
      handleSubmit(event);
    }
```


매우 많은 입력값을 유저로부터 받았는데, 유저가 확인할 시간도 없이, 바로 넘어가는건 좀 어색한 UX인거 같아요.

피그마에 있는 `다음` 버튼을 구현해보면 어떨까요?

bigsaigon333 commented on react-payments-app/src/components/Input/CardExpirationDate/input.jsx:
```diff
@@ -0,0 +1,74 @@
+  const [error, setError] = useState("");
+  const nextElement = useRef(null);
+
+  const setExpirationDateByIndex = (index) => (e) => {
```
set-* 함수는 호출하면 뭔가 상태를 set할 거 같은 뉘앙스인데, 
실제 이함수는 호출하면 상태를 set하는 함수를 반환하네요. 

다른 set-* 함수와는 결이 좀 다르기 때문에, 네이밍을 다른걸로 바꾸는게 좋을거 같아요!

bigsaigon333 commented on react-payments-app/src/components/Input/CardExpirationDate/input.jsx:
```diff
@@ -0,0 +1,74 @@
+    if (index === 0) {
+      if (Number(value) > VALID_INPUT.VALID_MONTH) {
+        setError("월은 1이상 12이하 숫자여야 합니다.");
+        expirationDate[index] = "";
```
❓ ❓ 
`setExpirationDate` 호출은 어디로..?

jlee0505 replied:
헉...수정하였습니다!

bigsaigon333 commented on react-payments-app/src/components/Input/CardExpirationDate/input.jsx:
```diff
@@ -0,0 +1,74 @@
+          expirationDate[index] = "";
+          return;
+        }
+      }
```
index로 분기가 왜 필요할까요? expirationDate가 배열이라서 그런가요?
그렇다면 expirationDate가 꼭 배열일 필요가 있을까요? 

한번 생각해보셨으면 좋겠습니다 😄

bigsaigon333 commented on react-payments-app/src/components/Input/CardSecurityCode/input.jsx:
```diff
@@ -0,0 +1,41 @@
+  const [CVC, setCVC] = useState("");
+  const [error, setError] = useState("");
+
+  const setCVCByInput = (e) => {
```
리액트에서 이런 핸들러함수들은 `onCVCChange`, `handleCVCChange` 등으로 명명하는게 일반적인 컨벤션인거 같습니다 😄 

핸들러 함수 컨벤션에 대해서는 갑론을박이 있으니 검색해보시는걸 추천드리고, 
하나의 컨벤션을 채택하게 되면 일관되게 다 써주시는게 중요한거 같아요!

jlee0505 replied:
그렇네요! onChange를 다루는 함수명들을 컴포넌트별로 다 다르게 쓰고 있었네요.

CardRegistraqtionPage에서는 handleChange,
Input들에서는 onChange나 set이름byInput으로 쓰던 것을 

모두 handleChange꼴로 통일시켜주었습니다!

bigsaigon333 commented on react-payments-app/src/components/Input/CardSecurityCode/input.jsx:
```diff
@@ -0,0 +1,41 @@
+import { MAX_INPUT_LENGTH } from "../../../common/constant";
+
+const CardCVCInput = ({ onChange }) => {
+  const [CVC, setCVC] = useState("");
+  const [error, setError] = useState("");
+
+  const setCVCByInput = (e) => {
+    let updatedCVC = CVC;
+    const { value } = e.target;
+
+    if (Number.isNaN(+value)) {
+      setError("숫자만 입력해주세요.");
+      return;
+    }
+
+    updatedCVC = value;
+    setCVC(updatedCVC);
+  };
+
+  useEffect(() => {
+    onChange(CVC, error);
+  }, [CVC, error]);
```
### src/components/Input/ 내 모든 디렉토리에 있는 Input 컴포넌트에 대한 리뷰입니다. 

CardOwner/input.jsx, CardPassword/input.jsx 등이 재사용가능할까요?
도메인과 매우 강하게 결합되어 때문에, 재사용이 어려운 일회용 컴포넌트입니다.
모든 컴포넌트가 도메인과 상관이 없는 재사용 가능한 컴포넌트 일수는 없기 때문에, 이는 자연스럽습니다.

그럼 components 디렉토리 아래에 있는 컴포넌트 중에 재사용가능한 컴포넌트가 있을까요?
카드 번호 입력이 아니라, 전화번호 입력 페이지를 추가하면, 어떤 컴포넌트를 재사용할 수 있을까요?
제가 생각하기에는 전화번호 입력 페이지에 필요한 모든 컴포넌트를 다시 만들어야 할 거 같아요.

input.jsx 내의 인풋 컴포넌트들은 비슷한 역할을 수행합니다.
상태가 있고, 상태를 변경하는 함수가 있고, 상태를 변경하기 전에 값을 검증하고, 에러가 있다면, 에러값을 상태로 설정하고...

이러한 부분들을 떼어내서 Input 컴포넌트를 공통 컴포넌트로 만들어낼 수 있지 않을까요? 
아니면 useInput 훅 같은걸 구현할 수 있지 않을까요?

---

페이먼츠 미션에서 말하는 CDD는, 도메인 로직을 구현해가다가 필요한 컴포넌트들이 생기면 그때그때 만들면서 가는 방식이 아니라,(Top-down)
도메인과 무관하게 유연하게 사용할 수 있는 컴포넌트들을 먼저 만들고, 이를 구성해서 도메인 로직을 완성시키는 방식입니다(Bottom-up)

사고방식을 완전히 바꾸셔야 합니다. 익숙하지 않으면 처음부터 요구사항만 보고 어떤 컴포넌트에 어떤 props를 받으면 되겠다를 떠올리기가 쉽지 않습니다. 하지만, 지금 전체적으로 도메인 로직을 다 구현해보셨기 때문에, 현재 구현한 Input 컴포넌트들이 필요로 하는 기능들만이라도 다 제공하는 그런 공통 Input 컴포넌트를 만들어보시기를 추천드립니다 😄 

아마 하다보면 고민되는 포인트가 몇개 있으실거에요. 
그 부분은 한번 현정님이 구현해보고 나서 다시 저랑 논의해봤으면 좋겠습니다 😺

bigsaigon333 commented on react-payments-app/src/components/Input/CardOwner/input.jsx:
```diff
@@ -0,0 +1,43 @@
+      return;
+    }
+  };
+  useEffect(() => {
+    onChange(cardOwner, error);
+  }, [cardOwner, error]);
```
CardForm에도 cardOwner가 있는데 CardOwnerInput에 cardOwner를 둘 필요가 있을까요? 
어떤 이점이 있고, 어떤 단점이 있을까요?

한번 고민해보셨으면 좋겠습니다 !

bigsaigon333 reviewed (changes requested):
현정님 안녕하세요~

페이먼츠 미션 Step1 구현하시라 고생 많으셨습니다~!
파일이 많아서 저도 생각보다 리뷰를 많이 남기게 되었네요 😅
보시면서 이해 안되는 부분 있다면 편하게 말씀주세요! 
(제 말씀을 잘못 이해한 채로 구현에 너무 많은 시간을 쓰시면 그건 그것대로 맘찢.. 😭 )

아래는 프로젝트 전반적 드리는 말씀입니다!

---


디렉토리 deps를 `react-payments-app` 로 하나더 하신 이유가 있으실까요? 
레포명으로 이미 페이먼츠 미션이라는게 드러나기 때문에, 굳이 라는 생각이 들었어요.
그리고 JS 패키지는 항상 최상단에 package.json이 있어야 하기 때문에, 이런 룰에도 맞지 않는다는 생각이 들었습니다 😄 

---

파일명 대소문자 이슈가 있네요!  clone 받아서 실행시켜보니 아래와 같은 에러가 나와서 제일 처음에 실행이 안되었어요. 

![image](https://user-images.githubusercontent.com/31029000/219868534-99a0a15a-a77f-48c9-878b-98268306bb4c.png)

이건 git 설정과도 관련이 있는데요, 찾아보시면 많이 나올겁니다 ㅎㅎ..

(새로 레포를 clone 받아서 실행시켜보면, 무슨 말씀인지 바로 이해되실겁니다~)


---

`npm run start` 했더니, lint warning 메세지가 많이 나오네요. 
warning 이라도 꼼꼼하게 보시고 다 수정하시는게 버그를 줄이는 길입니다. 

lint 관련 익스텐션도 설치하시고, commit 전에 자동으로 lint, prettier 를 돌려주는 그런 자동화 도구도 찾아보시면 재밌습니다 😄 

![image](https://user-images.githubusercontent.com/31029000/219868550-12953c8c-ef4d-48a3-8d18-cbf049033faa.png)

![image](https://user-images.githubusercontent.com/31029000/219869223-72cac5a5-6e30-4632-b1af-5abbc82d6052.png)


---

#### UI 관련 어색한 점

1. 에러메세지를 보고 다시 제대로 된 값을 입력하였는데도 계속해서 에러메시지가 남아있어, 혼란스러웠습니다

![image](https://user-images.githubusercontent.com/31029000/219869056-591efdb4-78a8-4ac3-97ed-3a15947ac853.png)

2. 이름을 입력하지 않았을 때 글자수가 없는 것과, 이름을 입력한 후 지워도 계속 1이 남아 있는 것 어색합니다.

![image](https://user-images.githubusercontent.com/31029000/219869123-ff4d1bdd-84ce-4a2d-aa47-4991808d7bb6.png)

![image](https://user-images.githubusercontent.com/31029000/219869109-57ae6416-63b9-461c-aefc-4b5660d20e4c.png)

jlee0505 committed:
initial commit

jlee0505 committed:
chore: lint 설치

-eslint
-prettier

jlee0505 committed:
chore: husky 설치 및 pre-commit 설정

jlee0505 committed:
chore: lint-staged 설치 및 husky pre-commit 재설정

jlee0505 committed:
fix: fix lint error

jlee0505 committed:
feat: ListPage -> RegistrationPage 이동기능

jlee0505 committed:
refactor: 카드 등록 페이지 -> 카드 등록 완료 페이지로 넘어가는 방식 변경

- 기존방식: 마지막 비밀번호를 입력하면, validation 진행 후 페이지 이동
- 고친방식: 사용자가 직접 다음 버튼을 눌러야 validation 및 페이지 이동 가능

jlee0505 committed:
feat: 카드UI에 실시간 정보 반영

jlee0505 committed:
fix: 변수명 변경으로 인한 오류 해결

jlee0505 committed:
fix: 초기 등록 카드 없을 때 렌더링 에러 임시처리

jlee0505 committed:
design: 실시간 카드UI 반영 카드번호 끝 2자리 *처리

jlee0505 committed:
design: wrapper 추가

jlee0505 committed:
refactor: 정확한 값을 입력하면 하단 에러메세지 사라지도록 설정

jlee0505 committed:
refactor: 카드 소유자 이름 없을 경우 0이 되도록 UI반영

jlee0505 committed:
Merge branch 'main' of https://github.com/jlee0505/react-payments into ver-2

jlee0505 committed:
1차 피드백 반영 (#4)

* chore: init project

- cra 설치
- storybook 설치
- 페이지 기반 라우팅
- 컴포넌트 분리 골격 잡기

* chore: storybook을 이용한 UI컴포넌트(정적버전) 구축

* chore: 누락된 Button UI 컴포넌트 추가

* chore: Header UI 컴포넌트 스타일 변경

* chore: class -> className 변경

* chore: UI컴포넌트 구조 수정

- 기존 RegistrationPage에서 CardNicknamePage 분리

* chore: 마이너 스타일 변경

* docs: step.1 기능 구현을 위한 구상 및 기능 목록 수정

* feat: 뒤로 가기 기능 추가

* feat: 카드 번호 입력 기능 추가

- CardForm 아래 input들은 모두 캡슐화했습니다.
-- 이유: CardRegistrationPage를 Header, Card, CardForm 으로 선언적 명시하기 위해서

* feat: 만료일 입력 기능 추가

* feat: 카드 소유자 입력 기능 추가

* feat: 보안코드(CVC) 입력 기능 추가

* feat: 비밀번호 입력 기능 추가

* feat: 데이터 전역관리를 위한 store 생성

* fix: 컨텍스트 에러 고침

* chore: 상수화

* feat: 입력을 마치면 카드 등록 완료 페이지로 이동

* initial commit

* chore: lint 설치

-eslint
-prettier

* chore: husky 설치 및 pre-commit 설정

* chore: lint-staged 설치 및 husky pre-commit 재설정

* fix: fix lint error

* feat: ListPage -> RegistrationPage 이동기능

* refactor: 카드 등록 페이지 -> 카드 등록 완료 페이지로 넘어가는 방식 변경

- 기존방식: 마지막 비밀번호를 입력하면, validation 진행 후 페이지 이동
- 고친방식: 사용자가 직접 다음 버튼을 눌러야 validation 및 페이지 이동 가능

* feat: 카드UI에 실시간 정보 반영

* fix: 변수명 변경으로 인한 오류 해결

* fix: 초기 등록 카드 없을 때 렌더링 에러 임시처리

* design: 실시간 카드UI 반영 카드번호 끝 2자리 *처리

* design: wrapper 추가

* refactor: 정확한 값을 입력하면 하단 에러메세지 사라지도록 설정

* refactor: 카드 소유자 이름 없을 경우 0이 되도록 UI반영

jlee0505 commented:
동희님 안녕하세요! 꼼꼼한 리뷰 감사드립니다. 정말 많은 도움이 되었어요 👍

주신 피드백을 크게 아래와 같이 정리해 반영해보았어요.

## 피드백 반영사항

### 1. 전반적인 에러사항 처리

[bcabb2d](https://github.com/next-step/react-payments/pull/43/commits/bcabb2d246f435a0029d23b9e3910bd521a958f0) 실수로 모든 변경사항이 요 커밋 하나에 들어가게 되었습니다. 큰 커밋 덩어리를 던져드려서 죄송해요🥲)

- 파일명 대소문자 이슈: .jsx 파일이 소문자로 시작해 생기는 에러를 수정
- 린트 설치 및 자동화: eslint, prettier, husky 를 설치 및 세팅, 설치 후 발생하는 에러를 수
- class 이슈: class→ className 수정
- 첫 화면 라우팅 이슈: + 버튼 누르면 라우팅이 되게끔 수정
- 디렉토리 이중구조 변경 이슈: 질문사항 참조 부탁드리겠습니다!

### 2. 어색한 UI display 개선

- 입력 오류 메세지 이슈: 부정확한 입력값을 정확히 수정해도 에러메세지가 남아있는 에러 수정
- 남은 입력 가능 이름 길이 이슈: 이름이 없을 때는 0으로 표시되도록 수정

### 3. 린트

- eol 이슈 → vscode설정, prettier 설정 완료
- 키워드, import/export, 변수명, 함수명에 일관적인 convention 적용

### 4. 고민해볼 점

- `<Header />` 변경에 취약한 구조 → 어떻게 하면 더 유연한 컴포넌트로 만들 수 있을까?
- `<CardExpirationDate />` 로직을 인덱스를 쓰지 않고 쓰는 법
- `useEffect` 의 지양: 언제? 무슨 이유로?
- 반복되는 `Input` 로직 → `reducer` 를 이요한 `useInput` 으로 만들어보기
- `cardOwner` 상태관리 → ???
- props type 관리 → TS의 사용, `knobs addon` 등의 사용 고려

위 4번 사항들은 더 깊게 고민해볼 문제인 것 같아 우선 다른 부분들을 먼저 고쳐 pr드렸습니다. 차후 미션을 진행하며 계속 고민하고 디벨롭 시켜나가 보겠습니다! 


## 추가사항

진행하다보니 눈에 띠는 개선사항들도 생겨 진행해보았습니다.

피드백 주신 부분은 아니었지만 작업하며 눈에 띄는 부분을 추가로 개선하였습니다.

- 정보 입력시 카드UI에 실시간으로 반영되도록 기능 구현
- Wrapper 추가


## 질문사항

위의 과정을 진행하며 고민해본 것들 중 이부분은 어떻게 처리해야하는 것이 좋을지 헷갈립니다 🤔

1. 클론 후 cra 설치시 폴더 구조 중복 피하는 법이 궁금합니다!
    
    > "디렉토리 deps를 react-payments-app 로 하나더 하신 이유가 있으실까요?
    레포명으로 이미 페이먼츠 미션이라는게 드러나기 때문에, 굳이 라는 생각이 들었어요.
    그리고 JS 패키지는 항상 최상단에 package.json이 있어야 하기 때문에, 이런 룰에도 맞지 않는다는 생각이 들었습니다 😄"
    
    굳이 depth를 추가한 다른 의도가 있던 것은 아니었어요! 클론을 받고 cra를 설치하니 자동으로 이런 구조가 만들어지게 되었습니다. 
    
    이후 피드백을 반영하며 수동으로 파일들을 옮겨 depth 를 같도록 만들어 주려 했으나, 이럴 경우 husky 폴더가 .git 과 같은 depth에 위치해주어야 한다는 오류를 접해 결국 husky 설치를 위해 이중 구조를 유지해야만 했습니다. **directory depth 를 맞추는 방법**이 있다면…궁금합니다!
    

감사합니다 🙇🏻‍♀️

jlee0505 committed:
chore: 불필요한 console.log 제거

jlee0505 committed:
chore: idex.html 수정

- 불필요한 테그들 삭제
- title 태그 변경

jlee0505 committed:
chore: Single story hoisting

jlee0505 committed:
chore: 지난 병합에서 잘못 합쳐진 파일들 삭제

jlee0505 committed:
chore: CardExpirationDateInput 작은 수정

- Month인풋 0이하 숫자 입력 가능한 오류 수정
- 에러시 상태 초기화 setState 이용하도록 수정

jlee0505 committed:
chore: event handler 이름 형식 통일

jlee0505 committed:
chore: css클래스 사이 개행문자 추가

- 가독성 향상을 위해 css 클래스 사이 개행문자 추가해 주었습니다.

jlee0505 committed:
chore: 불필요한 fragment 삭제

jlee0505 committed:
chore: import 컨벤션 통일

jlee0505 committed:
chore: end of line 자동화

jlee0505 committed:
chore: 불필요한 익명함수 콜백 삭제

jlee0505 committed:
Merge Commit

jlee0505 committed:
Merge pull request #5 from jlee0505/ver-2

1차 피드백 반영 (2)

bigsaigon333 commented on react-payments-app/.eslintrc.json:
```diff
@@ -0,0 +1,25 @@
+    "react/react-in-jsx-scope": "off",
+    "react/prop-types": "off",
+    "max-depth": 2,
+    // "max-lines": ["error", 15],
```
json 포맷에는 주석이라는 개념이 없습니다. 
따라서 이 파일은 적법한 json 형태가 아닙니다.

주석을 지우시거나, .eslintrc.json을 .eslintrc.js 등으로 변경하는 게 좋을거 같아요 👍🏻

bigsaigon333 reviewed (approved):
현정님 안녕하세요~

리뷰 반영하시느라 고생하셨습니다.

Step2 진행하면서, 남아있는 리뷰들과 CDD에 더 도전해보셨으면 좋겠습니다 😄 

Step2, 3도 남았으니 Step1은 여기서 approve 할게요~(계속 댓글 달아주셔도 좋아요 👍🏻 👍🏻 👍🏻 )

Step 2에서 만나요 💪🏻 💪🏻 💪🏻 


---

> 클론 후 cra 설치시 폴더 구조 중복 피하는 법
> > "디렉토리 deps를 react-payments-app 로 하나더 하신 이유가 있으실까요?
> > 레포명으로 이미 페이먼츠 미션이라는게 드러나기 때문에, 굳이 라는 생각이 들었어요.
>  > 그리고 JS 패키지는 항상 최상단에 package.json이 있어야 하기 때문에, 이런 룰에도 맞지 않는다는 생각이 들었습니다 😄"
> 
> 👉 사실 디렉토리의 중복구조를 피해주고 싶었어요... 하지만 클론을 받고 cra를 설치하니 자동으로 이런 구조가 만들어지게 되었습니다. 굳이 depth를 추가한 다른 의도가 있던 것은 아니었어요!
> 이후 피드백을 반영하며 수동으로 파일들을 옮겨 depth 를 같도록 만들어 주려 했으나, 이럴 경우 husky 폴더가 .git 과 같은 depth에 위치해주어야 한다는 오류룰 접해 결국 husky 설치를 위해 이중 구조를 유지해야만 했습니다. directory depth 를 어떻게 맞출 수 있을지 리뷰어님의 의견이 궁금합니다!

CRA를 설치할 때 어떤 디렉토리에 설치할지를 정할 수 있습니당~
(`npx create-react-app 디렉토리명`)

클론 받는 것과 만약 충돌이 된다면, CRA로 프로젝트를 먼저 생성한 다음에, 
해당 디렉토리에서 remote repository를 지정하는 방법도 있겠네요 😄 

그리고, 말씀하신 husky도 `.git`과 package.json이 같은 디렉토리에 안 있어도 사용하게 하는 방법이 있습니다 (모노레포 에서도 husky는 쓰이니까요~)

그런데, 지금은 husky와 .git 모두 최상위 디렉토리에 잘 위치하고 있어서, 
react-payments-app 안에 있는것만 꺼내면 잘 될것 같은데, 아닌가요~?

<img width="404" alt="image" src="https://user-images.githubusercontent.com/31029000/220223569-1685e0c5-bbd0-4f16-9e72-d6ba2892421e.png">

제가 해보니까 저는 잘 되는거 같아요 😄 

<img width="600" alt="image" src="https://user-images.githubusercontent.com/31029000/220225152-b038c779-b3b5-42bb-a6e6-2b36e73f3da0.png">

---

node_modules 디렉토리가 file changed에 올라와있네요!
2단계 진행하면서 삭제해주세요~ 👍🏻
